core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceTask.java

```diff
@@ -0,0 +1,93 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.sql.async;
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.sql.StatefulAtom;
+import io.questdb.std.DirectLongList;
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Misc;
+
+import java.io.Closeable;
+
+public class PageFrameReduceTask implements Closeable {
+    private final DirectLongList rows;
+    private final long pageFrameQueueCapacity;
+    private int frameIndex = Integer.MAX_VALUE;
+    private PageFrameSequence<?> frameSequence;
+
+    public PageFrameReduceTask(CairoConfiguration configuration, int pageFrameQueueCapacity) {
+        this.rows = new DirectLongList(configuration.getPageFrameReduceRowIdListCapacity(), MemoryTag.NATIVE_LONG_LIST);
+        this.pageFrameQueueCapacity = pageFrameQueueCapacity;
+    }
+
+    @Override
+    public void close() {
+        Misc.free(rows);
+    }
+
+    public void collected() {
+        final long frameCount = frameSequence.getFrameCount();
+        // We have to reset capacity only on max all queue items
+        // What we are avoiding here is resetting capacity on 1000 frames given our queue size
+        // is 32 items. If our particular producer resizes queue items to 10x of the initial size
+        // we let these sizes stick until produce starts to wind down.
+        if (frameIndex >= frameCount - pageFrameQueueCapacity) {
```

Comment:
nit: might be a good idea to force `resetCapacity()` if current capacity exceeds a threshold (similar to how it's done with `CompiledFilterRecordCursor#rowsCapacityThreshold`). This would help us to preserve a soft limit for the row id lists' memory footprint.

Reply:
Marking this one as resolved since the max page frame count is now 1M by default.